dune/grid/common/ZoltanGraphFunctions.cpp

```diff
@@ -188,6 +230,130 @@ void getCpGridEdgeList(void *cpGridPointer, int sizeGID, int sizeLID,
+    {
+        *err = ZOLTAN_FATAL;
+        return;
+    }
+    int oldidx = 0;
+    int idx = 0;
+
+    for( int cell = 0; cell < numCells;  cell++ )
+    {
+        const int currentCell = localID[cell];
+
+        // First the strong edges of the well completions.
+        auto wellEdges = graph.getWellsGraph()[currentCell];
+        well_perf += wellEdges.size();
+        for( auto edge : wellEdges)
+        {
+            nborGID[idx] = edge;
+            ewgts[idx++] = std::numeric_limits<float>::max();
+        }
+
+        // Now the ones of the grid that are not handled by the well completions
+        for ( int local_face = 0 ; local_face < grid.numCellFaces(static_cast<int>(localID[cell])); ++local_face )
+        {
+            const int face  = grid.cellFace(currentCell, local_face);
+            int otherCell   = grid.faceCell(face, 0);
+            if ( otherCell == currentCell || otherCell == -1 )
+            {
+                otherCell = grid.faceCell(face, 1);
+                if ( otherCell == currentCell || otherCell == -1 )
+                {
+                    // no real face or already handled by well
+                    continue;
+                }
+                else
+                {
+                    if ( wellEdges.find(otherCell) == wellEdges.end() )
+                    {
+                        nborGID[idx] = globalID[otherCell];
+                        ewgts[idx++] = 1;
+                    }
+                    continue;
+                }
+            }
+            if ( wellEdges.find(otherCell) == wellEdges.end() )
+            {
+                nborGID[idx] = globalID[otherCell];
+                ewgts[idx++] = 1;
+            }
+        }
+        assert(idx-oldidx==numEdges[cell]);
+        oldidx = idx;
+    }
+
+    const int myrank = grid.comm().rank();
+
+    for ( int i = 0; i < idx; ++i )
+    {
+        nborProc[i] = myrank;
+    }
+#if defined(DEBUG) && false // The index set will not be initialized here!
+    // The above relies heavily on the grid not being distributed already.
+    // Therefore we check here that all cells are owned by us.
+    GlobalLookupIndexSet<Dune::CpGrid::ParallelIndexSet>
+        globalIdxSet(grid.getCellIndexSet(),
+                     grid.numCells());
+    for ( int cell = 0; cell < numCells;  cell++ )
+    {
+        if ( globalIdxSet.pair(cell)->local().attribute() !=
+             Dune::CpGrid::ParallelIndexSet::LocalIndex::Attribute::owner )
+        {
+            *err = ZOLTAN_FATAL;
+        }
+    }
+#endif
+}
+
+CombinedGridWellGraph::CombinedGridWellGraph(const CpGrid& grid,
+                      const Opm::EclipseStateConstPtr eclipseState)
+    : grid_(grid)
+{
+    wellsGraph_.resize(grid.numCells());
+    // create compressed lookup from cartesian.
+    std::vector<Opm::WellConstPtr> wells  = eclipseState->getSchedule()->getWells();
+    std::vector<int> cartesian_to_compressed(grid.numCells(), -1);
```

Comment:
That's almost certainly a programming error.  Unless something's changed very recently, the `numCells` method returns the number of active (compressed) cells but you need the number of global cells here (i.e., the product of the `cartDims()`).